source/cppfront.cpp

```diff
@@ -1828,9 +1776,60 @@ class cppfront
                 }
             }
 
+            // Going backwards if we found LeftParen it might be UFCS
+            // expr_list is emited to args variable for future use
+            if (i->op->type() == lexeme::LeftParen) {
+
+                args.emplace();
+
+                if (!i->expr_list->expressions.empty()) {
+                    args.emplace(print_to_text_chunks(*i->expr_list));
+                } 
+                
+            }
+            // Going backwards if we found Dot and there is args variable
+            // it means that it should be handled by UFCS
+            else if( i->op->type() == lexeme::Dot && args )
+            {
+                auto funcname = print_to_string(*i->id_expr);
+
+                //--------------------------------------------------------------------
+                //  TODO: When MSVC supports __VA_OPT__ in standard mode without the
+                //        experimental /Zc:preprocessor switch, use this single line
+                //        instead of the dual lines below that special-case _0 args
+                //  AND:  Make the similarly noted change in cpp2util.h
+                //
+                //printer.print_cpp2("CPP2_UFCS(", n.position());
+
+                auto ufcs_string = std::string("CPP2_UFCS");
+
+                if (i->id_expr->template_args_count() > 0) {
+                    ufcs_string += "_TEMPLATE";
+                    // we need to replace "fun<int,long,double>" to "fun, (<int,long,double>)" to be able to generate
+                    // from obj.fun<int, long, double>(1,2) this CPP2_UFCS_TEMPLATE(fun, (<int,long, double>), obj, 1, 2)
+                    auto split = funcname.find('<'); assert(split != std::string::npos);
+                    funcname.insert(split, ", (");
+                    funcname += ')';
+                }
+                //  If there are no additional arguments, use the _0 version
+                if (args.value().empty()) {
+                    ufcs_string += "_0";
+                }
+
+                prefix.emplace_back(ufcs_string + "(" + funcname + ", ", i->op->position() );
+                suffix.emplace_back(")", i->op->position() );
+                if (!args.value().empty()) {
+                    for (auto&& e: args.value()) {
+                        suffix.push_back(e);
+                    }
+                    suffix.emplace_back(", ", i->op->position());
```

Comment:
(Aside: This position isn't precise but is probably fine because we're rewriting so much, and this position is probably earlier than we're currently at which means it won't move anything. There are other places, particularly in this function, where I don't use a precise position for similar reasons.)